### DISCUSSION TITLE
Custom components crashes with state functionality

### DISCUSSION THREAD

**Discussion by chadcwilliams** [AutoResearch/autora#545] [Using AutoRA]
I am working on part 4 of the tutorials, which is making custom AutoRA components. I am not too sure what is required for theorists/experiment runners/experimentalists to function with the wrappers. I tried what worked with the collector but that didn't work out (see below). I couldn't find any documentation on it, but likely looking in the wrong places. 

----------------------------

I create a state:
```
#### Import Modules ####
import numpy as np
import pandas as pd
import torch
from autora.variable import DV, IV, ValueType, VariableCollection
from autora.state.bundled import StandardState
from autora.state.delta import on_state
from autora.state.wrapper import state_fn_from_estimator
from autora.experimentalist.random_ import random_pool, random_sample

#### Define metadata ####
iv = IV(name="x", value_range=(0, 2 * np.pi), allowed_values=np.linspace(0, 2 * np.pi, 10))
dv = DV(name="y", type=ValueType.REAL)
variables = VariableCollection(independent_variables=[iv],dependent_variables=[dv])

conditions = random_pool(variables, num_samples=10)

#### Initiate state ####
s_poly = StandardState(variables = variables, conditions = conditions, experiment_data = pd.DataFrame(columns=["x","y"]))
```

I add an experiment runner and experimentalist:
```
#### Define experiment runner and push to state ####
def run_experiment(conditions: pd.DataFrame):
    x = conditions["x"]
    y = np.sin(x) + np.random.normal(0, 0.5, size=x.shape)
    observations = conditions.assign(y = y)
    return observations

experiment_runner = on_state(run_experiment, output=["experiment_data"])

experimentalist = on_state(random_sample, output=["conditions"])
```

I have the following custom theorist:

```import numpy as np
from sklearn.base import BaseEstimator

class PolynomialRegressor(BaseEstimator):

    def __init__(self, degree: int = 3):
        self.degree = degree

    def fit(self, conditions, observations):

        # polyfit expects a 1D array
        if conditions.ndim > 1:
            conditions = conditions.flatten()

        if observations.ndim > 1:
            observations = observations.flatten()

        # fit polynomial
        self.coeff = np.polyfit(conditions, observations, self.degree)
        self.polynomial = np.poly1d(self.coeff)
        pass

    def predict(self, conditions):
        return self.polynomial(conditions)

theorist = state_fn_from_estimator(PolynomialRegressor())
```

However, when I run:
```
s_poly = theorist(experiment_runner(experimentalist(s_poly )))
```

I get the following error:
```
AttributeError                            Traceback (most recent call last)
[~\AppData\Local\Temp\ipykernel_41048\951453664.py](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/cwill/GitHub/autora/docs/tutorials/basic/~/AppData/Local/Temp/ipykernel_41048/951453664.py) in ?()
      1 print('\033[1mPrevious State:\033[0m')
      2 print(s_poly)
      3 
      4 for cycle in range(5):
----> 5     s_poly = theorist(experiment_runner(experimentalist(s_poly)))
      6 
      7 print('\n\033[1mUpdated State:\033[0m')
      8 print(s_poly)

[c:\Users\cwill\GitHub\virtualEnvs\autoraEnv\lib\site-packages\autora\state\delta.py](file:///C:/Users/cwill/GitHub/virtualEnvs/autoraEnv/lib/site-packages/autora/state/delta.py) in ?(state_, **kwargs)
   1031     @wraps(f)
   1032     def _f(state_: S, **kwargs) -> S:
-> 1033         delta = f(state_, **kwargs)
   1034         assert isinstance(delta, Mapping), (
   1035             "Output of %s must be a `Delta`, `UserDict`, " "or `dict`." % f
   1036         )

[c:\Users\cwill\GitHub\virtualEnvs\autoraEnv\lib\site-packages\autora\state\delta.py](file:///C:/Users/cwill/GitHub/virtualEnvs/autoraEnv/lib/site-packages/autora/state/delta.py) in ?(state_, **kwargs)
    764                 align_dataframe_to_ivs(arguments["conditions"],
    765                                        getattr(state_, "variables").independent_variables)
    766             )
    767 
...
   5987         ):
   5988             return self[name]
-> 5989         return object.__getattribute__(self, name)

AttributeError: 'DataFrame' object has no attribute 'flatten'
```

This occurs with non-theorist components as well.

**Comment by younesStrittmatter** [comment]
Hard to tell. At first glance, I see that your fit function doesn't have a return value but should return self. I am not sure if this is causing the error though.

**Comment by younesStrittmatter** [comment]
This happens because conditions and observations in your regressor fit function are pd DataFrames. They can not be flattened. This should work:
`from sklearn.base import BaseEstimator

class PolynomialRegressor(BaseEstimator):

    def __init__(self, degree: int = 3):
        self.degree = degree

    def fit(self, conditions, observations):
        c = np.array(conditions)
        o = np.array(observations)

        # polyfit expects a 1D array
        if conditions.ndim > 1:
            c = c.flatten()

        if observations.ndim > 1:
            o = o.flatten()

        # fit polynomial
        self.coeff = np.polyfit(c, o, self.degree)
        self.polynomial = np.poly1d(self.coeff)
        pass

    def predict(self, conditions):
        c = np.array(conditions)
        return self.polynomial(c)`

**Reply by younesStrittmatter** [reply]
ups. Forgot to replace something: 

`class PolynomialRegressor(BaseEstimator):

    def __init__(self, degree: int = 3):
        self.degree = degree

    def fit(self, conditions, observations):
        c = np.array(conditions)
        o = np.array(observations)

        # polyfit expects a 1D array
        if c.ndim > 1:
            c = c.flatten()

        if o.ndim > 1:
            o = o.flatten()

        # fit polynomial
        self.coeff = np.polyfit(c, o, self.degree)
        self.polynomial = np.poly1d(self.coeff)
        pass

    def predict(self, conditions):
        c = np.array(conditions)
        return self.polynomial(c)`

**Comment by chadcwilliams** [comment]
Thanks Younes, that worked! 

Simple enough fix, but tough to determine from the error code.
